sql/src/main/java/io/crate/executor/transport/task/UpsertByIdTask.java

```diff
@@ -55,10 +73,99 @@ public UpsertByIdTask(UUID jobId,
                           TransportCreateIndexAction transportCreateIndexAction,
                           UpsertByIdNode node) {
         super(jobId);
-
+        this.transportShardUpsertActionDelegate = transportShardUpsertActionDelegate;
+        this.transportCreateIndexAction = transportCreateIndexAction;
+        this.clusterService = clusterService;
         this.node = node;
+        autoCreateIndex = new AutoCreateIndex(settings);
+
+        if (node.items().size() == 1) {
+            // skip useless usage of bulk processor if only 1 item in statement
+            // and instead create upsert request directly on start()
+            resultList = new ArrayList<>(1);
+            resultList.add(SettableFuture.<TaskResult>create());
+        } else {
+            resultList = initializeBulkShardProcessor(settings);
+        }
+
+    }
+
+    @Override
+    public void start() {
+        if (node.items().size() == 1) {
+            // directly execute upsert request without usage of bulk processor
+            @SuppressWarnings("unchecked")
+            final SettableFuture<TaskResult> futureResult = (SettableFuture)resultList.get(0);
+            final UpsertByIdNode.Item item = node.items().get(0);
+            if (node.isPartitionedTable()) {
+                createIndexIfRequired(item.index(), futureResult);
```

Comment:
if it is just a single insert/update it would be possible to chain the futures instead of blocking on the index creation